auth/src/main/java/org/aerogear/auth/impl/UserPrincipalImpl.java

```diff
@@ -85,13 +85,14 @@ Builder withEmail(final String email) {
         }
 
         Builder withRole(final IRole role) {
-            this.roles.put(role.getRoleName(), role);
+            RoleKey roleKey = new RoleKey(role);
+            this.roles.put(roleKey, role);
             return this;
         }
 
-        Builder withRoles(final IRole[] roles) {
+        Builder withRoles(final Map<RoleKey, IRole> roles) {
```

Comment:
This should take an array of roles. The `Map` should be populated internally. The `RoleKey` object is not to be seen by the caller.